### ISSUE TITLE
minimind项目中GRPO函数实现时，为什么优势函数的新旧概率比是这样的

### ISSUE THREAD

**Issue by yangxuan14nlp** [jingyaogong/minimind#563]
<img width="1318" height="736" alt="Image" src="https://github.com/user-attachments/assets/ade271b4-adf2-40d0-9a6f-c0ebf740b325" />

        per_token_logps = get_per_token_logps(model, outputs, completion_ids.size(1))  # [B*num_gen, R]
        with torch.no_grad():
            ref_per_token_logps = get_per_token_logps(ref_model, outputs, completion_ids.size(1))  # [B*num_gen, R]

        completions = tokenizer.batch_decode(completion_ids, skip_special_tokens=True)
        rewards = calculate_rewards(prompts, completions, reward_model, reward_tokenizer).to(args.device)  # [B*num_gen]

        grouped_rewards = rewards.view(-1, args.num_generations)  # [B, num_gen]
        mean_r = grouped_rewards.mean(dim=1).repeat_interleave(args.num_generations)  # [B*num_gen]
        std_r = grouped_rewards.std(dim=1).repeat_interleave(args.num_generations)  # [B*num_gen]
        advantages = torch.clamp((rewards - mean_r) / (std_r + 1e-4), -10, 10)
        advantages = (advantages - advantages.mean()) / (advantages.std() + 1e-8)  # [B*num_gen]

        is_eos = completion_ids == tokenizer.eos_token_id  # [B*num_gen, R]
        eos_idx = torch.full((is_eos.size(0),), is_eos.size(1), dtype=torch.long, device=args.device)
        eos_idx[is_eos.any(dim=1)] = is_eos.int().argmax(dim=1)[is_eos.any(dim=1)]
        completion_mask = (torch.arange(is_eos.size(1), device=args.device).expand(is_eos.size(0), -1) <= eos_idx.unsqueeze(1)).int()  # [B*num_gen, R]

        kl_div = ref_per_token_logps - per_token_logps
        per_token_kl = torch.exp(kl_div) - kl_div - 1  # [B*num_gen, R]
        per_token_loss = -(torch.exp(per_token_logps - per_token_logps.detach()) * advantages.unsqueeze(1) - args.beta * per_token_kl)  # [B*num_gen, R]
        loss = ((per_token_loss * completion_mask).sum(dim=1) / completion_mask.sum(dim=1)).mean() / args.accumulation_steps  # scalar
        loss.backward()

请看这行：     per_token_loss = -(torch.exp(per_token_logps - per_token_logps.detach()) * advantages.unsqueeze(1) - args.beta * per_token_kl)  # [B*num_gen, R]
这行都是per_token_logps， 只是第二个多了一个 detach()，  按照我的理解值应该是一样的吧，比值不为1了吗？

我查了一下知乎上的：https://zhuanlan.zhihu.com/p/1889782947894233018
        advantages = inputs["advantages"]

        # 重要性采样per_token_logps.detach()对应旧策略的sample
        old_per_token_logps = inputs["old_per_token_logps"] if self.num_iterations > 1 else per_token_logps.detach()

        # torch.exp(per_token_logps - old_per_token_logps) 等价于 e^(log per1 - log per2)，即e^log per1 / e^log per2
        # 相当于per_token_logps /old_per_token_logps, 即新的策略/旧的策略.
        # coef_1 等价于per_token_logits/old_per_token_logits，也就是重要性采样
        coef_1 = torch.exp(per_token_logps - old_per_token_logps)
        # torch.clamp对应公式中的梯度裁剪CLIP
        coef_2 = torch.clamp(coef_1, 1 - self.epsilon_low, 1 + self.epsilon_high)
        per_token_loss1 = coef_1 * advantages.unsqueeze(1)
        per_token_loss2 = coef_2 * advantages.unsqueeze(1)
        
        # 最终计算loss的时候会在前面取负号“-”
        # 原因在于通过梯度上升求A的最大值等价于通过梯度下降求-A的最小值。
        per_token_loss = -torch.min(per_token_loss1, per_token_loss2)
        if self.beta != 0.0:
            per_token_loss = per_token_loss + self.beta * per_token_kl
        loss = (per_token_loss * completion_mask).sum() / completion_mask.sum()
        return loss
他这里感觉有新旧比值不为1的体现，

这让我比较疑惑，热切期望有大佬解答

**Comment by jingyaogong** [comment]
数值上确实恒等于1，但关键要看梯度怎么算。

设 $l = \log\pi_\theta$：

- `exp(0)` → 前向=1，反向梯度=0
- `exp(l - l.detach())` → 前向=1，反向梯度= $\frac{\partial l}{\partial \theta}$ 

**1、传统REINFORCE写法**：

$$loss_1 = -A \cdot l$$

```python
loss = -advantages * log_probs
```

$$\frac{\partial loss_1}{\partial \theta} = -A \cdot \frac{\partial l}{\partial \theta}$$

**2、这里的写法**：

$$loss_2 = -A \cdot e^{l - l_{stop}}$$

其中 $l_{stop}$ 是 `l.detach()`，常数且数值就 = $l$

因为 $\frac{\partial l_{stop}}{\partial \theta} = 0$，且 $l - l_{stop} = 0$

所以 

$$\frac{\partial loss_2}{\partial \theta} = -A \cdot e^{l - l_{stop}} \cdot \frac{\partial (l - l_{stop})}{\partial \theta} = -A \cdot e^{0} \cdot \frac{\partial l}{\partial \theta} = -A \cdot \frac{\partial l}{\partial \theta}$$

**两者梯度完全一样。**

---

### 为什么要这么写？

形式上更接近PPO/GRPO论文里的 importance sampling ratio $\frac{\pi_{new}}{\pi_{old}}$，即 `exp(log_new - log_old)`。

好处是方便扩展：如果以后要改成多次迭代（off-policy），只需把 `l.detach()` 换成保存的 `old_logps`，改动最小（你提到的知乎 `old_per_token_logps` 就是）

```python
old_per_token_logps = inputs.get("old_per_token_logps")
old_per_token_logps = per_token_logps.detach() if old_per_token_logps is None else old_per_token_logps

log_ratio = per_token_logps - old_per_token_logps
coef_1 = torch.exp(log_ratio)
```

这里用的是单次on-policy，每次生成完直接更新，old和new是同一个策略，所以直接用 `detach()` 就行；btw，TRL GRPO Trainer默认是单次的
